File: mobile/src/main/java/org/openhab/habdroid/ui/activity/ContentControllerTwoPane.java
/*
 * Copyright (c) 2018, openHAB.org and others.
 *
 *   All rights reserved. This program and the accompanying materials
 *   are made available under the terms of the Eclipse Public License v1.0
 *   which accompanies this distribution, and is available at
 *   http://www.eclipse.org/legal/epl-v10.html
 */

package org.openhab.habdroid.ui.activity;

import android.os.Bundle;
import android.util.Pair;
import android.view.View;
import android.view.ViewStub;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import org.openhab.habdroid.R;
import org.openhab.habdroid.model.LinkedPage;
import org.openhab.habdroid.ui.MainActivity;
import org.openhab.habdroid.ui.WidgetListFragment;

@SuppressWarnings("unused") // instantiated via reflection
public class ContentControllerTwoPane extends ContentController {
    private View mRightContentView;

    public ContentControllerTwoPane(MainActivity activity) {
        super(activity);
    }

    @Override
    public void onRestoreInstanceState(Bundle state) {
        super.onRestoreInstanceState(state);
        boolean rightPaneVisible = mFm.findFragmentById(R.id.content_right) != null;
        mRightContentView.setVisibility(rightPaneVisible ? View.VISIBLE : View.GONE);
    }

    @Override
    protected void executeStateUpdate(FragmentUpdateReason reason, boolean allowStateLoss) {
        Fragment leftFragment = getOverridingFragment();
        final WidgetListFragment rightFragment;
        final Pair<LinkedPage, WidgetListFragment> rightPair;

        if (leftFragment != null) {
            rightFragment = null;
            rightPair = null;
        } else if (mSitemapFragment != null) {
            rightPair = mPageStack.empty() ? null : mPageStack.peek();
            leftFragment = mPageStack.size() > 1
                    ? mPageStack.get(mPageStack.size() - 2).second : mSitemapFragment;
            rightFragment = rightPair != null ? rightPair.second : null;
        } else {
            leftFragment = mDefaultProgressFragment;
            rightFragment = null;
            rightPair = null;
        }

        Fragment currentLeftFragment = mFm.findFragmentById(R.id.content_left);
        Fragment currentRightFragment = mFm.findFragmentById(R.id.content_right);

        FragmentTransaction removeTransaction = mFm.beginTransaction();
        boolean needRemove = false;
        if (currentLeftFragment != null && currentLeftFragment != leftFragment) {
            removeTransaction.remove(currentLeftFragment);
            needRemove = true;
        }
        if (currentRightFragment != null && currentRightFragment != rightFragment) {
            removeTransaction.remove(currentRightFragment);
            needRemove = true;
        }
        if (needRemove) {
            if (allowStateLoss) {
                removeTransaction.commitNowAllowingStateLoss();
            } else {
                removeTransaction.commitNow();
            }
        }

        FragmentTransaction ft = mFm.beginTransaction();
        ft.setCustomAnimations(R.anim.slide_in_left, R.anim.slide_out_right);
        if (leftFragment != null) {
            ft.setCustomAnimations(determineEnterAnim(reason), determineExitAnim(reason));
            ft.replace(R.id.content_left, leftFragment);
            if (leftFragment instanceof WidgetListFragment) {
                WidgetListFragment llf = (WidgetListFragment) leftFragment;
                llf.setHighlightedPageLink(rightPair != null ? rightPair.first.link() : null);
            }
        }
        if (rightFragment != null) {
            ft.setCustomAnimations(0, 0);
            ft.replace(R.id.content_right, rightFragment);
            rightFragment.setHighlightedPageLink(null);
        }
        if (allowStateLoss) {
            ft.commitAllowingStateLoss();
        } else {
            ft.commit();
        }

        mRightContentView.setVisibility(rightFragment != null ? View.VISIBLE : View.GONE);
    }

    @Override
    public void openPage(LinkedPage page, WidgetListFragment source) {
        Fragment currentLeftFragment = mFm.findFragmentById(R.id.content_left);
        if (source == currentLeftFragment && !mPageStack.empty()) {
            mPageStack.pop();
        }
        super.openPage(page, source);
    }

    @Override
    protected WidgetListFragment getFragmentForTitle() {
        return mPageStack.size() > 1
                ? mPageStack.get(mPageStack.size() - 2).second
                : mSitemapFragment;
    }

    @Override
    public void inflateViews(ViewStub stub) {
        stub.setLayoutResource(R.layout.content_twopane);
        View view = stub.inflate();
        mRightContentView = view.findViewById(R.id.content_right);
        mRightContentView.setVisibility(View.GONE);
    }
}

File: mobile/src/main/java/org/openhab/habdroid/ui/activity/ContentControllerTwoPane.kt
/*
 * Copyright (c) 2010-2020 Contributors to the openHAB project
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package org.openhab.habdroid.ui.activity

import android.os.Bundle
import android.view.View
import android.view.ViewStub
import androidx.core.view.isVisible
import androidx.fragment.app.commit
import androidx.fragment.app.commitNow
import org.openhab.habdroid.R
import org.openhab.habdroid.model.LinkedPage
import org.openhab.habdroid.ui.MainActivity
import org.openhab.habdroid.ui.WidgetListFragment

// instantiated via reflection
@Suppress("UNUSED")
class ContentControllerTwoPane(activity: MainActivity) : ContentController(activity) {
    private lateinit var rightContentView: View

    override val fragmentForTitle
        get() = if (pageStack.size > 1) pageStack[pageStack.size - 2].second else sitemapFragment

    override fun onRestoreInstanceState(state: Bundle) {
        super.onRestoreInstanceState(state)
        rightContentView.isVisible = fm.findFragmentById(R.id.content_right) != null
    }

    override fun executeStateUpdate(reason: FragmentUpdateReason, allowStateLoss: Boolean) {
        var leftFragment = overridingFragment
        val rightFragment: WidgetListFragment?
        val rightPair: Pair<LinkedPage, WidgetListFragment>?

        when {
            leftFragment != null -> {
                rightFragment = null
                rightPair = null
            }
            sitemapFragment != null -> {
                rightPair = if (pageStack.empty()) null else pageStack.peek()
                leftFragment = fragmentForTitle
                rightFragment = rightPair?.second
            }
            else -> {
                leftFragment = defaultProgressFragment
                rightFragment = null
                rightPair = null
            }
        }

        val currentLeftFragment = fm.findFragmentById(R.id.content_left)
        val currentRightFragment = fm.findFragmentById(R.id.content_right)

        fm.commitNow(allowStateLoss) {
            if (currentLeftFragment != null && currentLeftFragment !== leftFragment) {
                remove(currentLeftFragment)
            }
            if (currentRightFragment != null && currentRightFragment !== rightFragment) {
                remove(currentRightFragment)
            }
        }

        fm.commit(allowStateLoss) {
            setCustomAnimations(R.anim.slide_in_left, R.anim.slide_out_right)
            if (leftFragment != null) {
                setCustomAnimations(determineEnterAnim(reason), determineExitAnim(reason))
                replace(R.id.content_left, leftFragment)
                if (leftFragment is WidgetListFragment) {
                    leftFragment.setHighlightedPageLink(rightPair?.first?.link)
                }
            }
            if (rightFragment != null) {
                setCustomAnimations(0, 0)
                replace(R.id.content_right, rightFragment)
                rightFragment.setHighlightedPageLink(null)
            }
        }

        rightContentView.isVisible = rightFragment != null
    }

    override fun openPage(page: LinkedPage, source: WidgetListFragment) {
        val currentLeftFragment = fm.findFragmentById(R.id.content_left)
        if (source === currentLeftFragment && !pageStack.empty()) {
            pageStack.pop()
        }
        super.openPage(page, source)
    }

    override fun inflateViews(stub: ViewStub) {
        stub.layoutResource = R.layout.content_twopane
        val view = stub.inflate()
        rightContentView = view.findViewById(R.id.content_right)
        rightContentView.isVisible = false
    }
}
